adapter/kubernetes/kubernetes.go

```diff
@@ -184,23 +193,30 @@ func (*builder) ValidateConfig(c adapter.Config) (ce *adapter.ConfigErrors) {
 }
 
 func (b *builder) BuildAttributesGenerator(env adapter.Env, c adapter.Config) (adapter.AttributesGenerator, error) {
-	var clientErr error
 	paramsProto := c.(*config.Params)
-	b.once.Do(func() {
+	b.Lock()
+	defer b.Unlock()
+	if b.needsCacheInit {
 		refresh := paramsProto.CacheRefreshDuration
-		controller, err := b.newCacheControllerFn(paramsProto.KubeconfigPath, refresh, env)
+		path, exists := os.LookupEnv("KUBECONFIG")
+		if !exists {
+			path = paramsProto.KubeconfigPath
+		}
+		controller, err := b.newCacheControllerFn(path, refresh, env)
```

Comment:
this happens in the first line of `newCacheControllerFn`:

`env.Logger().Infof("getting kubeconfig from: %#v", kubeconfigPath)`